packages/sink-mongo/src/collection.ts

```diff
@@ -0,0 +1,184 @@
+import type { Cursor } from "@apibara/protocol";
+import type {
+  BulkWriteOptions,
+  ClientSession,
+  Collection,
+  Condition,
+  DeleteOptions,
+  Document,
+  Filter,
+  FindCursor,
+  FindOptions,
+  InsertManyResult,
+  InsertOneOptions,
+  InsertOneResult,
+  MatchKeysAndValues,
+  OptionalUnlessRequiredId,
+  UpdateFilter,
+  UpdateOptions,
+  UpdateResult,
+  WithId,
+} from "mongodb";
+
+export type MongoCursor = {
+  from: number | null;
+  to: number | null;
+};
+
+export type CursoredSchema<TSchema extends Document> = TSchema & {
+  _cursor: MongoCursor;
+};
+
+export class MongoSinkCollection<TSchema extends Document> {
+  constructor(
+    private session: ClientSession,
+    private collection: Collection<TSchema>,
+    private endCursor?: Cursor,
+  ) {}
+
+  async insertOne(
+    doc: OptionalUnlessRequiredId<TSchema>,
+    options?: InsertOneOptions,
+  ): Promise<InsertOneResult<TSchema>> {
+    return await this.collection.insertOne(
+      {
+        ...doc,
+        _cursor: {
+          from: Number(this.endCursor?.orderKey),
+          to: null,
+        } as MongoCursor,
+      },
+      { ...options, session: this.session },
+    );
+  }
+
+  async insertMany(
+    docs: ReadonlyArray<OptionalUnlessRequiredId<TSchema>>,
+    options?: BulkWriteOptions,
+  ): Promise<InsertManyResult<TSchema>> {
+    return await this.collection.insertMany(
+      docs.map((doc) => ({
+        ...doc,
+        _cursor: {
+          from: Number(this.endCursor?.orderKey),
+          to: null,
+        } as MongoCursor,
+      })),
+      { ...options, session: this.session },
+    );
+  }
+
+  async updateOne(
+    filter: Filter<TSchema>,
+    update: UpdateFilter<TSchema>,
+    options?: UpdateOptions,
+  ): Promise<UpdateResult<TSchema>> {
+    return await this.collection.updateOne(
+      {
+        ...filter,
+        _cursor: {
+          to: null,
+        } as Condition<MongoCursor | null>,
+      },
+      {
+        ...update,
+        $set: {
+          ...update.$set,
+          "_cursor.to": Number(this.endCursor?.orderKey),
+        } as unknown as MatchKeysAndValues<TSchema>,
+      },
+      { ...options, session: this.session },
+    );
```

Comment:
I believe this one should also duplicate the previously existing value. The easiest way (double check to make sure I'm correct) is to use `findOneAndUpdate` asking to return the value _before_ the update. Notice that in this case the update `$set` action should set the lower bound to the endCursor (since we are updating in place the old value to be the new value).

 - `const updated = this.collection.findOneAndUpdate(..., ..., { returnDocument: "before" })`
 - `updated._cursor[to] = endCursor.orderKey`
 - `this.collection.insert(updated)`;